wurst/network/Network.wurst

```diff
@@ -0,0 +1,922 @@
+		integers, reals, booleans, strings and serializables.
+		Because we want to keep gamecache keys below a certain size, we can't store
+		everything immediately in the gamecache without risking exhausting available keys.
+		We also need to know the size of data being sent prior to starting the transmission,
+		so we have to store all of it in an intermediate buffer, which is DataStream.
+		Prior to sending, all strings in the DataStream are 'exploded' into a stream
+		of integers, because SyncStoredString doesn't work.
+		After sending, they are 'compacted' back into strings and written to the DataStream.
+
+		GamecacheStream is a gamecache-based container with stream semantics for writing
+		integers, reals and booleans. There is a GamecacheStream for each primitive type,
+		integer, boolean, real and stringInts.
+
+		Network is the main class that coordinates DataStream and GamecacheStream and does
+		all the heavy lifting.
+
+		Before starting the transmission, the DataStream is locked into a non-writable, non-readable
+		mode to prevent anyone from trying to mess with it while the transmission is going.
+		Strings are also exploded into ints.
+		The maximum amount of data across all primitive stream is calculated, and the amount of 
+		required 'sync rounds' is calculated - that is, the amount of times we need to flush/sync
+		data out of the gamecaches to keep key sizes short.
+		We use a SimpleSynchronizer to know when all metadata has been sent and received.
+
+		Since only the local player has any knowledge about the amount of data needed to be sent,
+		and consequently, the amount of sync rounds required, we first send a pre-fetch "metadata"
+		payload with the amount of data in each stream and the amount of sync rounds, using fixed
+		keys.
+
+		When all players receive the metadata, we can start syncing each round. We flush the caches,
+		load up to gcKeysCount values into each GamecacheStream from the DataStream, and then sync it,
+		and use SimpleSynchronizer to know when all data is available for reading.
+
+		When a round is received, we write data to the DataStream, using the metadata to know when
+		to stop, and start another sync round if necessary. If it is not necessary,
+		we open the DataStream for reading and call the finish callback, and destroy the instance.
+********************************************************************************/
+
+// characters that can be sent over the network
+constant string STRING_ALPHABET = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+// characters that can be used as GC keys
+// their amount also determines the size of a single 'round'
+constant string GC_KEYS  = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`{|}~"
+
+// precomputed set of GC keys
+string array gcKeys
+// count of GC keys
+integer gcKeysCount = 0
+// lookup table for alphabet
+Table alphabetLookup
```

Comment:
Should be a constant, and can be a HashMap ?

Reply:
Would introduce useless overhead (key string stored in typecasting hashtable and translated back/forth to be used in another table) for slightly nicer syntax. Not worth it in a library imo.

Reply:
HashMap provides compiletime safety - safety is never useless overhead.

But that doesn't mean it's the correct choice 100% of the time...

Reply:
You only use save/load/has int. You shouldn't need typecasting at all and thus there will be no overhead?!

Reply:
Not true, I also store strings and reals?